### PR TITLE
Support spaces in S3 paths

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:

**Bug fix**

## Description:

- S3 paths allow spaces.
- Adjusting URI Parsing to allow for spaces (they are escaped by AWS SDK before sending it)
- Fixes https://github.com/NVIDIA/DALI/issues/5521

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [x] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/util/uri.h [hunk 1]
```diff
@@ -47,7 +47,28 @@ class URI {
...
+    AllowNonEscaped = 1 << 0,  // 0x0001 - Convenient when we want to check if a URI is valid before
+                               // escaping it (e.g. replacing spaces with %20).
+    // Add more options as needed
+  };
+
+  /**
+   * @brief
+   *
+   * @param uri URI to parse
+   * @param allow_non_escaped whether we should allow for non escaped characters in path, query and
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/util/uri.h]: Fixed

File: dali/util/uri.h [hunk 2]
```diff
@@ -47,7 +47,25 @@ class URI {
...
+  enum ParseOpts : uint32_t {
+    Default = 0,
+    AllowNonEscaped = 1 << 0,  // 0x0001 - Convenient when we want to check if a URI is valid before
+                               // escaping it (e.g. replacing spaces with %20).
+    // Add more options as needed
+  };
+
+  /**
+   * @brief
+   *
```
Comment (mzient) [inline dali/util/uri.h:63]: Remove or add description?

File: dali/util/uri.h [hunk 3]
```diff
@@ -47,7 +47,25 @@ class URI {
   }
 
  public:
-  static DLL_PUBLIC URI Parse(std::string uri);
+  /**
+   * @brief Parse options
+   * 
+   */
```
Comment (mzient) [inline dali/util/uri.h:53]: Nitpick:
```suggestion
  /** Parse options */
```
^^^ This is also a "brief" description, but takes much fewer lines.

File: dali/operators/reader/loader/indexed_file_loader.h [hunk 4]
```diff
@@ -171,8 +169,8 @@ class IndexedFileLoader : public Loader<CPUBackend, Tensor<CPUBackend>, true> {
       } else {
         tensor.Resize({size}, DALI_UINT8);
 
-        int64 n_read = current_file_->Read(reinterpret_cast<uint8_t*>(tensor.raw_mutable_data()),
-                            size);
+        int64 n_read =
+            current_file_->Read(reinterpret_cast<uint8_t*>(tensor.raw_mutable_data()), size);
```
Comment (mzient) [inline dali/operators/reader/loader/indexed_file_loader.h:173]:
```suggestion
            current_file_->Read(static_cast<uint8_t*>(tensor.raw_mutable_data()), size);
```
...while you're at it.

**Review by banasraf — APPROVED** [review]

**Review by mzient — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**